### DISCUSSION TITLE
What is the "extraction" syntax in transformation functions called?

### DISCUSSION THREAD

**Discussion by andrewjroberts** [pointfreeco/swift-composable-architecture#1978] [Q&A] [answered]
Hi all,

In the VoiceMemos example (and elsewhere), mapping/transformation functions such as `fromChildAction` use syntax like the following:
```swift
action: /Action.voiceMemo(id:action:)
```

- What is the Swift language feature that enables us to capture the id:action: using that smooshed sort of syntax?
- What is this syntax actually doing behind the scenes if it's sugar?

Cheers!

**Comment by tgrapperon** [comment, marked as answer]
Hey @andwrobs! There are various things at play here, but you should have a look at [`CasePaths`](https://github.com/pointfreeco/swift-case-paths) for more informations about the inner bits.

A `CasePath` is something similar to a `KeyPath` that is able to extract/embed associated values for a given enum's case.

You can form a `CasePath` using explicit `extract` and `embed` implementations, but the library is also able to perform some reflection magic and to create a complete `CasePath` from a function `(AssociatedValue) -> Enum`.

Enum's cases with associated values are promoted to static functions of this exact shape when used without argument. For example:
```swift
enum E {
  case s(String)
}
```
`E.s` is inferred as a function `(String) -> E`, because `E.s evaluated with ("Something")` creates the `E.s("something")` enum's value (I don't know the corresponding Swift Evolution reference).

So what you see in your example is a custom prefix operator. Namely `/`. This operator expects a `(UUID, VoiceMemo.Action) -> Action` suffix function, and you pass `Action.voiceMemo(id:action:)`.
It creates the `CasePath` corresponding to this enum's case. In other words:
```swift
[ / + (UUID, VoiceMemo.Action) -> Action ] ~> CasePath.init(embed: (UUID, VoiceMemo.Action) -> Action) ~> CasePath<Action, VoiceMemo.Action>
```

I don't know if this is clear! There is indeed a bunch of medium/high level features involved for this to work, but `CasePath`'s source is relatively short and you can probably get a better idea by looking at it (you don't need to understand the reflection implementation BTW! (I don't!))

**Reply by andrewjroberts** [reply]
clear top to bottom, helps connect a lot of the dots, and gives me plenty to lookup to learn more 🙏 

ty ty for explaining @tgrapperon !
